### PR TITLE
[Merged by Bors] - feat(Analysis): integer complement in `ℂ`

### PR DESCRIPTION
**Pull request by CBirkbeck**
Develop some basic API for the complement of the integers in the complex numbers.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Analysis/Complex/IntegerCompl.lean [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+Authors: Chris Birkbeck
+-/
+
+import Mathlib.Analysis.Complex.UpperHalfPlane.Basic
+
+/-!
+# Integer Complement
+
+We define the complement of the integers in the complex plane and give some basic lemmas about it.
+We also show that the upper half plane is mapped to the integer complement.
```
Comment (b-mehta) [inline Mathlib/Analysis/Complex/IntegerCompl.lean:13]:
```suggestion
We also show that the upper half plane embeds into the integer complement.
```
perhaps?

File: Mathlib/Analysis/Complex/IntegerCompl.lean [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+theorem Complex.closedEmbedding_coe_complex : ClosedEmbedding ((↑) : ℤ → ℂ) := by
+  apply Metric.closedEmbedding_of_pairwise_le_dist zero_lt_one
+  convert Int.pairwise_one_le_dist
+  simp_rw [dist_eq_norm]
+  norm_cast
+  rw [Int.norm_eq_abs]
+  exact Int.cast_abs
+
+lemma ℂ_ℤ_Isclosed : IsClosed (((↑) : ℤ → ℂ)'' ⊤) := by
```
Comment (b-mehta) [inline Mathlib/Analysis/Complex/IntegerCompl.lean:32]:
```suggestion
lemma ℂ_ℤ_Isclosed : IsClosed (((↑) : ℤ → ℂ) '' ⊤) := by
```

File: Mathlib/Analysis/Complex/IntegerCompl.lean [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+  refine IsClosed.not ?_
+  convert ℂ_ℤ_Isclosed
+  ext y
+  aesop
+
+instance : LocallyCompactSpace ℂ_ℤ := IsOpen.locallyCompactSpace ℂ_ℤ_IsOpen
+
+instance : Coe ℂ_ℤ ℂ := ⟨fun x => x.1⟩
+
+lemma upper_half_plane_ne_int (z : ℍ) : ∀ n : ℤ, z.1 ≠ n := by
```
Comment (Ruben-VandeVelde) [inline Mathlib/Analysis/Complex/IntegerCompl.lean:46]:
```suggestion
lemma UpperHalfPlane.ne_int (z : ℍ) : ∀ n : ℤ, z.1 ≠ n := by
```

File: Mathlib/Analysis/Complex/IntegerCompl.lean [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+  intro n
+  have h1 := z.2
+  aesop
+
+instance coe_upp : Coe ℍ ℂ_ℤ := ⟨fun x => ⟨x, by simpa using upper_half_plane_ne_int x⟩⟩
```
Comment (Ruben-VandeVelde) [inline Mathlib/Analysis/Complex/IntegerCompl.lean:51]:
```suggestion
instance : Coe ℍ ℂ_ℤ := ⟨fun x => ⟨x, by simpa using upper_half_plane_ne_int x⟩⟩
```

File: Mathlib/Analysis/Complex/IntegerCompl.lean [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+
+We define the complement of the integers in the complex plane and give some basic lemmas about it.
+We also show that the upper half plane is mapped to the integer complement.
+
+-/
+
+open UpperHalfPlane
+
+/--The complement of the integers in `ℂ`. -/
+def ℂ_ℤ := {z : ℂ // ¬ ∃ (n : ℤ), z = ↑n}
```
Comment (Ruben-VandeVelde) [inline Mathlib/Analysis/Complex/IntegerCompl.lean:20]: Can we give this a name and make this a (scoped) notation? Then you could use it in the names below as well

File: Mathlib/Analysis/Complex/IntegerCompl.lean [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+open UpperHalfPlane
+
+/--The complement of the integers in `ℂ`. -/
+def ComplexIntegerComplement := {z : ℂ // ¬ ∃ (n : ℤ), z = ↑n}
+
+namespace ComplexIntegerComplement
+
+local notation "ℂ_ℤ " => ComplexIntegerComplement
+
+noncomputable instance : UniformSpace ℂ_ℤ  :=  instUniformSpaceSubtype
```
Comment (Ruben-VandeVelde) [inline Mathlib/Analysis/Complex/IntegerCompl.lean:26]:
```suggestion
noncomputable instance : UniformSpace ℂ_ℤ := instUniformSpaceSubtype
```

File: Mathlib/Analysis/Complex/IntegerCompl.lean [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+
+theorem Complex.closedEmbedding_coe_complex : ClosedEmbedding ((↑) : ℤ → ℂ) := by
```
Comment (Ruben-VandeVelde) [inline Mathlib/Analysis/Complex/IntegerCompl.lean:28]: Please move to `Mathlib.Analysis.Complex.Basic` as `Complex.closedEmbedding_intCast`

File: Mathlib/Analysis/Complex/IntegerCompl.lean [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+  apply Metric.closedEmbedding_of_pairwise_le_dist zero_lt_one
+  convert Int.pairwise_one_le_dist
+  simp_rw [dist_eq_norm]
+  norm_cast
+  rw [Int.norm_eq_abs]
+  exact Int.cast_abs
+
+lemma ℂ_ℤ_Isclosed : IsClosed (((↑) : ℤ → ℂ) '' ⊤) := by
```
Comment (Ruben-VandeVelde) [inline Mathlib/Analysis/Complex/IntegerCompl.lean:36]: Please don't use top for sets, but also it seems like this should use `Set.range` directly. Also, this should move to `Mathlib.Analysis.Complex.Basic` as `Complex.isClosed_range_intCast`

File: Mathlib/Analysis/Complex/IntegerCompl.lean [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+  simp only [Set.top_eq_univ, Set.image_univ]
+  exact Complex.closedEmbedding_coe_complex.isClosed_range
+
+lemma ℂ_ℤ_IsOpen : IsOpen {z : ℂ | ¬ ∃ (n : ℤ), z = ↑n} := by
```
Comment (Ruben-VandeVelde) [inline Mathlib/Analysis/Complex/IntegerCompl.lean:40]: Also move, and I think state as
```lean
lemma Complex.isOpen_compl_range_intCast : IsOpen (Set.range ((↑) : ℤ → ℂ))ᶜ :=
  Complex.isClosed_range_intCast.isOpen_compl 
```

**Review by j-loreaux — COMMENTED** [review]
You'll need to merge master, as some things like `ClosedEmbedding` have been renamed.

**Comment by CBirkbeck** [comment]
So I changed the definition to now make it a `Set ℂ` instead of its own type. Originally I was just following the UpperHalfPlane API, but as @loefflerd points out, this is probably not needed here. But if anyone has a good reason why it should be its own type then I'd be interested to know!

File: Mathlib/Analysis/Complex/IntegerCompl.lean [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+open UpperHalfPlane
+
+/--The complement of the integers in `ℂ`. -/
+abbrev Complex.IntegerComplement := {z : ℂ | ¬ ∃ (n : ℤ), z = ↑n}
+
+namespace Complex
+
+local notation "ℂ_ℤ " => IntegerComplement
+
+noncomputable instance : UniformSpace ℂ_ℤ := instUniformSpaceSubtype
```
Comment (loefflerd) [inline Mathlib/Analysis/Complex/IntegerCompl.lean:26]: This is not needed with the `Set` version

File: Mathlib/Analysis/Complex/IntegerCompl.lean [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+
+instance : LocallyCompactSpace ℂ_ℤ := by
+  apply IsOpen.locallyCompactSpace
+  convert Complex.isOpen_compl_range_intCast
+  rw [IntegerComplement]
+  aesop
```
Comment (loefflerd) [inline Mathlib/Analysis/Complex/IntegerCompl.lean:32]: Nor this – better to have a lemma `isOpen_integerComplement`

File: Mathlib/Analysis/Complex/IntegerCompl.lean [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+
+/--The coercion from the upper half plane into the IntegerComplement. -/
+@[coe] def ucoe : ℍ → ℂ_ℤ := fun z => ⟨z, by simpa using UpperHalfPlane.ne_int z⟩
+
+instance : Coe ℍ ℂ_ℤ := ⟨ucoe⟩
```
Comment (loefflerd) [inline Mathlib/Analysis/Complex/IntegerCompl.lean:37]: Keeping track of coercions is already a serious headache in the modular forms code; let's not add even more of them. Why not just `lemma UpperHalfPlane.coe_mem_integerComplement (z : ℍ) : \u z \in ℂ_ℤ`?

File: Mathlib/Analysis/Complex/IntegerCompl.lean [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+  rw [add_eq_zero_iff_eq_neg] at h
+  have := not_exists.mp x.2 (-a)
+  aesop
+
+lemma IntegerComplement_ne_zero (x : ℂ_ℤ) : x.1 ≠ 0 := by
+  simpa using IntegerComplement_add_ne_zero x 0
+
+lemma IntegerComplement_ne_one (x : ℂ_ℤ) : x.1 ≠ 1 := by
+  have := IntegerComplement_add_ne_zero x (-1 : ℤ)
+  aesop
```
Comment (loefflerd) [inline Mathlib/Analysis/Complex/IntegerCompl.lean:50]: I would write the hypotheses as `{x : ℂ} (hx : x \in ℂ_ℤ)` rather than using a subtype

**Comment by CBirkbeck** [comment]
Ok I've added the suggestions from the reviews and I also changed the definition so actually be `(Set.range ((↑) : ℤ → ℂ))ᶜ` since that is what the instances are written as. I added some basic lemmas to relate to the previous definition. But I'm happy to change it back if people think this isn't ideal.

File: Mathlib/Analysis/Complex/IntegerCompl.lean [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+open UpperHalfPlane
+
+/--The complement of the integers in `ℂ`. -/
+abbrev Complex.IntegerComplement := (Set.range ((↑) : ℤ → ℂ))ᶜ
+
+namespace Complex
+
+local notation "ℂ_ℤ " => IntegerComplement
+
+instance : IsOpen ℂ_ℤ := isOpen_compl_range_intCast
```
[the remark this answers is not in the document]
Comment (CBirkbeck) [reply Mathlib/Analysis/Complex/IntegerCompl.lean]: No it wasnt supposed to be there. I was just testing if apply? would find the result, but forgot to delete it!

File: Mathlib/Analysis/Complex/IntegerCompl.lean [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+
+We define the complement of the integers in the complex plane and give some basic lemmas about it.
+We also show that the upper half plane embeds into the integer complement.
+
+-/
+
+open UpperHalfPlane
+
+/--The complement of the integers in `ℂ`. -/
+abbrev Complex.IntegerComplement := (Set.range ((↑) : ℤ → ℂ))ᶜ
```
Comment (urkud) [inline Mathlib/Analysis/Complex/IntegerCompl.lean:20]: This is a set, so it should use `lowerCamelCase`.
```suggestion
abbrev Complex.integerComplement := (Set.range ((↑) : ℤ → ℂ))ᶜ
```
BTW, why do you want to have it reducible?

File: Mathlib/Analysis/Complex/IntegerCompl.lean [hunk 16]
```diff
@@ -0,0 +1,55 @@
...
+
+namespace Complex
+
+local notation "ℂ_ℤ " => IntegerComplement
+
+lemma IntegerComplement_eq : ℂ_ℤ = {z : ℂ | ¬ ∃ (n : ℤ), z = ↑n} := by
+  rw [IntegerComplement]
+  aesop
```
Comment (j-loreaux) [inline Mathlib/Analysis/Complex/IntegerCompl.lean:28]: If you turn the equality around (which you *should* do here and elsewhere) this is just `rfl`:
```suggestion
lemma IntegerComplement_eq : ℂ_ℤ = {z : ℂ | ¬ ∃ (n : ℤ), n = z} := rfl
```

File: Mathlib/Analysis/Complex/IntegerCompl.lean [hunk 17]
```diff
@@ -0,0 +1,55 @@
...
+
+lemma IntegerComplement_not_exist {x : ℂ} (hx : x ∈ ℂ_ℤ) : ¬ ∃ (n : ℤ), x = ↑n := by
+  rw [IntegerComplement_eq] at hx
+  exact hx
+
+lemma IntegerComplement_mk {x : ℂ} (hx : ¬ ∃ (n : ℤ), x = ↑n) : x ∈ ℂ_ℤ := by
+  rw [IntegerComplement_eq]
+  exact hx
```
Comment (j-loreaux) [inline Mathlib/Analysis/Complex/IntegerCompl.lean:36]: Likewise:
```suggestion
lemma IntegerComplement_not_exist {x : ℂ} (hx : x ∈ ℂ_ℤ) : ¬ ∃ (n : ℤ), n = x := hx

lemma IntegerComplement_mk {x : ℂ} (hx : ¬ ∃ (n : ℤ), n = x) : x ∈ ℂ_ℤ := hx
```
But actually, the latter is misnamed now that it's not a subtype. I would probably just turn this into a `rw` lemma like this:
```
lemma integerComplement.mem_iff {x : ℂ} : x ∈ ℂ_ℤ ↔ ¬ ∃ (n : ℤ), n = x := Iff.rfl
```
and then you can add `alias`es if you prefer.

File: Mathlib/Analysis/Complex/IntegerCompl.lean [hunk 18]
```diff
@@ -0,0 +1,55 @@
...
+
+lemma UpperHalfPlane.coe_mem_integerComplement (z : ℍ) : ↑z ∈ ℂ_ℤ := by
+ rw [IntegerComplement_eq]
+ simpa using UpperHalfPlane.ne_int z
+
+lemma IntegerComplement_add_ne_zero {x : ℂ} (hx : x ∈ ℂ_ℤ) (a : ℤ) : x + (a : ℂ)  ≠ 0 := by
```
Comment (j-loreaux) [inline Mathlib/Analysis/Complex/IntegerCompl.lean:42]: You can have a stronger result here:
```suggestion
lemma integerComplement.add_coe_int_mem {x : ℂ} (a : ℤ) : x + (a : ℂ) ∈ ℂ_ℤ ↔ x ∈ ℂ_ℤ := by
```
And then your original result follows easily from this and the next result.

File: Mathlib/Analysis/Complex/IntegerCompl.lean [hunk 19]
```diff
@@ -0,0 +1,55 @@
...
+  intro h
+  rw [add_eq_zero_iff_eq_neg] at h
+  have := not_exists.mp hx (-a)
+  aesop
+
+lemma IntegerComplement_ne_zero {x : ℂ} (hx : x ∈ ℂ_ℤ) : x ≠ 0 := by
+  simpa using IntegerComplement_add_ne_zero hx 0
```
Comment (j-loreaux) [inline Mathlib/Analysis/Complex/IntegerCompl.lean:49]: I would namespace this, and the new proof avoids the previous lemma.
```suggestion
lemma integerComplement.ne_zero {x : ℂ} (hx : x ∈ ℂ_ℤ) : x ≠ 0 :=
  fun hx' ↦ hx ⟨0, by exact_mod_cast hx'.symm⟩
```

File: Mathlib/Analysis/Complex/IntegerCompl.lean [hunk 20]
```diff
@@ -0,0 +1,55 @@
...
+
+lemma IntegerComplement_ne_one {x : ℂ} (hx : x ∈ ℂ_ℤ): x ≠ 1 := by
+  have := IntegerComplement_add_ne_zero hx (-1 : ℤ)
+  aesop
```
Comment (j-loreaux) [inline Mathlib/Analysis/Complex/IntegerCompl.lean:53]: likewise:
```suggestion
lemma integerComplement.ne_one {x : ℂ} (hx : x ∈ ℂ_ℤ): x ≠ 1 :=
  fun hx' ↦ hx ⟨1, by exact_mod_cast hx'.symm⟩
```

File: Mathlib/Analysis/Complex/IntegerCompl.lean (hunk 18 above)
Comment (CBirkbeck) [reply Mathlib/Analysis/Complex/IntegerCompl.lean]: Sure! although my proof of this is not very pretty, maybe you have some ideas how to golf it? (also is there a contrapose for iffs??)

File: Mathlib/Analysis/Complex/IntegerCompl.lean [hunk 21]
```diff
@@ -0,0 +1,57 @@
...
+local notation "ℂ_ℤ " => integerComplement
+
+lemma IntegerComplement_eq : ℂ_ℤ = {z : ℂ | ¬ ∃ (n : ℤ), n = z} := rfl
+
+lemma integerComplement.mem_iff {x : ℂ} : x ∈ ℂ_ℤ ↔ ¬ ∃ (n : ℤ), n = x := Iff.rfl
+
+lemma UpperHalfPlane.coe_mem_integerComplement (z : ℍ) : ↑z ∈ ℂ_ℤ := by
+ rw [integerComplement.mem_iff]
+ simp only [not_exists]
+ exact fun x a ↦ (UpperHalfPlane.ne_int z)  x (id (Eq.symm a))
```
Comment (j-loreaux) [inline Mathlib/Analysis/Complex/IntegerCompl.lean:33]:
```suggestion
lemma UpperHalfPlane.coe_mem_integerComplement (z : ℍ) : ↑z ∈ ℂ_ℤ :=
  not_exists.mpr fun x hx ↦ ne_int z x hx.symm 
```

File: Mathlib/Analysis/Complex/IntegerCompl.lean [hunk 22]
```diff
@@ -0,0 +1,57 @@
...
+  · contrapose
+    simp only [Set.mem_compl_iff, Set.mem_range, not_exists, not_forall, Decidable.not_not,
+      forall_exists_index]
+    intro b hb
+    refine ⟨b + a, by simp only [Int.cast_add, hb]⟩
+  · contrapose
+    simp only [Set.mem_compl_iff, Set.mem_range, not_exists, not_forall, Decidable.not_not,
+      forall_exists_index]
+    intro b hb
+    refine ⟨b - a, by simp only [Int.cast_sub, hb, add_sub_cancel_right]⟩
```
Comment (j-loreaux) [inline Mathlib/Analysis/Complex/IntegerCompl.lean:46]:
```suggestion
  simp only [mem_iff, not_iff_not]
  exact ⟨(Exists.elim · fun n hn ↦ ⟨n - a, by simp [hn]⟩),
    (Exists.elim · fun n hn ↦ ⟨n + a, by simp [hn]⟩)⟩
```
or if you prefer it not so heavily golfed:
```suggestion
  simp only [mem_iff, not_iff_not]
  constructor
  · rintro ⟨n, hn⟩
    exact ⟨n - a, by simp [hn]⟩
  · rintro ⟨n, rfl⟩
    exact ⟨n + a, by simp⟩
```

File: Mathlib/Analysis/Complex/UpperHalfPlane/Basic.lean [hunk 23]
```diff
@@ -163,6 +163,16 @@ theorem normSq_ne_zero (z : ℍ) : Complex.normSq (z : ℂ) ≠ 0 :=
 theorem im_inv_neg_coe_pos (z : ℍ) : 0 < (-z : ℂ)⁻¹.im := by
   simpa using div_pos z.property (normSq_pos z)
 
+lemma ne_nat (z : ℍ) : ∀ n : ℕ, z.1 ≠ n := by
+  intro n
+  have h1 := z.2
+  aesop
+
+lemma ne_int (z : ℍ) : ∀ n : ℤ, z.1 ≠ n := by
+  intro n
```
Comment (j-loreaux) [inline Mathlib/Analysis/Complex/UpperHalfPlane/Basic.lean:172]: The proof of `UpperHalfPlane.coe_mem_integerComplement` mildly suggests that these lemmas should be turned around, but I'm not going to push for this.

**Review by j-loreaux — COMMENTED** [review]
Thanks! I've supplied some alternate proofs for you.

File: Mathlib/Analysis/Complex/IntegerCompl.lean (hunk 15 above)
Comment (j-loreaux) [reply Mathlib/Analysis/Complex/IntegerCompl.lean]: It should find it if you `rw [Complex.integerComplement]` and then use `apply?`. I suspect you don't want this to be an `abbrev`, but I could be wrong.

File: Mathlib/Analysis/Complex/IntegerCompl.lean (hunk 18 above)
Comment (j-loreaux) [reply Mathlib/Analysis/Complex/IntegerCompl.lean]: Done in another suggestion.

File: Mathlib/Analysis/Complex/IntegerCompl.lean (hunk 15 above)
Comment (CBirkbeck) [reply Mathlib/Analysis/Complex/IntegerCompl.lean]: Ah yes ok that works. I changed it back to a `def`. But out of curiosity, why do you think we don't want it to be an abbrev? (I dont have much of a feel for these things).

File: Mathlib/Analysis/Complex/UpperHalfPlane/Basic.lean (hunk 23 above)
Comment (CBirkbeck) [reply Mathlib/Analysis/Complex/UpperHalfPlane/Basic.lean:172]: Hmm sure, I'll see if it becomes annoying later on with the use case I have for this, if so I'll come back and change them.

File: Mathlib/Analysis/Complex/IntegerCompl.lean (hunk 15 above)
Comment (j-loreaux) [reply Mathlib/Analysis/Complex/IntegerCompl.lean]: I'm not totally sure that it should be `def` over `abbrev`; it's just my intuition here. I don't think there's anything strong pointing in either direction.

File: Mathlib/Analysis/Complex/IntegerCompl.lean [hunk 24]
```diff
@@ -0,0 +1,47 @@
...
+open UpperHalfPlane
+
+/--The complement of the integers in `ℂ`. -/
+def Complex.integerComplement := (Set.range ((↑) : ℤ → ℂ))ᶜ
+
+namespace Complex
+
+local notation "ℂ_ℤ " => integerComplement
+
+lemma IntegerComplement_eq : ℂ_ℤ = {z : ℂ | ¬ ∃ (n : ℤ), n = z} := rfl
```
Comment (j-loreaux) [inline Mathlib/Analysis/Complex/IntegerCompl.lean:26]:
```suggestion
lemma integerComplement_eq : ℂ_ℤ = {z : ℂ | ¬ ∃ (n : ℤ), n = z} := rfl
```

**Review by j-loreaux — COMMENTED** [review]
bors d+

**Pull request closed without merging**